File: 07/js/main.js
const fs = require('fs')
const {join} = require('path')

const file = fs.readFileSync(join(__dirname, '../input.txt'), 'utf8')
const inputs = file.split(',').map(n => Number(n))

// Problem 1
{
  let record = Infinity

  for (let i = 0; i < 1000; i++) {
    let sum = 0
    inputs.forEach(num => {
      sum += Math.abs(i - num)
    })

    if (sum < record) {
      record = sum
    }
  }

  console.log(`Answer 1: ${record}`);
}


// Problem 2
{
  function calc(num) {
    return num / 2 * num + num / 2
  }

  let record = Infinity

  for (let i = 0; i < 1000; i++) {
    let sum = 0
    inputs.forEach(num => {
      sum += calc(Math.abs(i - num))
    })

    if (sum < record) {
      record = sum
    }
  }

  console.log(`Answer 2: ${record}`);
}

File: 07/c/main.c
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <stdbool.h>

#define N 1024
int crabs[N] = {0};
int crabs_size = 0;

int abs(int num) {
  if (num > 0) return num;
  return -num;
}

int calc(int num) {
  float fnum = (float) num;
  float res = fnum / 2 * fnum + fnum / 2;
  return (int)res;
}

int main(int argc, char *argv[]) {
  assert(argc == 2 && "Must provide one argument => input.txt path");

  printf("Input file: %s\n\n", argv[1]);
  FILE *file = fopen(argv[1], "r");

  int number;
  while (fscanf(file, "%d", &number) == 1) {
    crabs[crabs_size++] = number;
    fgetc(file);
  }

  // Problem 1
  {
    long long record = 99999999;

    for (int i = 0; i < 1000; i++) {
      long long sum = 0;
      for (int j = 0; j < crabs_size; j++) {
        sum += abs(i - crabs[j]);
      }

      if (sum < record) {
        record = sum;
      }
    }

    printf("Answer 1: %lld\n", record);
  }


  // Problem 2
  {
    long long record = 99999999;

    for (int i = 0; i < 1000; i++) {
      long long sum = 0;
      for (int j = 0; j < crabs_size; j++) {
        sum += calc(abs(i - crabs[j]));
      }

      if (sum < record) {
        record = sum;
      }
    }

    printf("Answer 2: %lld\n", record);
  }


  fclose(file);

  return 0;
}
